### PR TITLE
Bug 1211395 - Performance regression in Message

### PR DESCRIPTION
**Pull request by steveck-chung**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=79c96b02c697a73b3d6712e6c1b5e3cf5ceda6c5

File: apps/sms/views/conversation/js/conversation.js [hunk 1]
```diff
@@ -1844,8 +1843,17 @@ var ConversationView = {
       this.mainWrapper.classList.toggle('edit');
     }
 
+    this.editHeader.removeAttribute('no-font-fit');
+
     if (!this.selectionHandler) {
-      LazyLoader.load('/views/shared/js/selection_handler.js', () => {
+      LazyLoader.load([
```
Comment (steveck-chung) [inline apps/sms/views/conversation/js/conversation.js]: Looks like this part could be wrapped into utils. BTW it's amazing that we have 3 different edit-mode.css for an edit form...

File: apps/sms/views/shared/js/navigation.js [hunk 2]
```diff
@@ -880,6 +881,10 @@ var Navigation = {
    */
   setReady() {
     debug('setReady()');
+
+    Array.forEach(document.querySelectorAll('.panel-hidden'), (panel) => {
```
Comment (julienw) [inline apps/sms/views/shared/js/navigation.js]: nit: Array.forEach is not standard, please use Array.from(...).forEach()

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6a1ba3e8860c090f8608d1233ebd58b1ab9d2843

File: apps/sms/views/shared/js/startup.js [hunk 3]
```diff
@@ -79,8 +79,23 @@ var Startup = exports.Startup = {
...
+    '/views/conversation/style/attachment.css',
+    '/views/conversation/style/conversation.css',
+    '/views/new_message/style/new_message.css',
+    '/views/new_message/style/recipients.css'
+  ],
+
   _lazyLoadInit: function() {
-    var lazyLoadPromise = LazyLoader.load(this._lazyLoadScripts).then(() => {
+    var lazyLoadFiles = this._lazyLoadScripts.concat(this._lazyLoadStyles);
+    var lazyLoadPromise = LazyLoader.load(lazyLoadFiles).then(() => {
```
Comment (azasypkin) [inline apps/sms/views/shared/js/startup.js]: __optional nit:__ you probably can do ```load([...this._lazyLoadScripts, ...this._lazyLoadStyles])```

File: apps/sms/index.html [hunk 4]
```diff
@@ -134,7 +134,7 @@
   <body role="application" class="skin-comms theme-communications">
     <iframe src="/views/shared/shim_host.html" class="shim-host" hidden></iframe>
     <article id="main-wrapper" class="wrapper">
-      <section class="panel panel-InboxView" role="region" aria-hidden="true">
+      <section class="panel panel-InboxView panel-hidden" role="region" aria-hidden="true">
```
Comment (azasypkin) [inline apps/sms/index.html]: __question:__ just curious, what part of the perf win is brought by introducing of ```.panel-hidden```?

Comment (steveck-chung) [inline apps/sms/index.html]: `.panel-hidden` is display: none actually that make sure we do nothing for the panels which are not necessary to be displayed first.

Comment (azasypkin) [inline apps/sms/index.html]: Yeah, I see, I just was wondering how much exactly this change decreases visuallyLoaded time comparing to the rest of the changes :)

Comment (steveck-chung) [inline apps/sms/index.html]: Some rough result in https://bugzilla.mozilla.org/show_bug.cgi?id=1211395#c17

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f7d5bdfa68868e3044fcd5ac92399235440c0592

File: apps/sms/index.html (hunk 4 above)
Comment (azasypkin) [inline apps/sms/index.html]: Oh, 100ms, great!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ec870306e8c9f6caf30b2281925eadba314b7551

File: apps/sms/views/shared/test/unit/navigation_test.js [hunk 5]
```diff
@@ -611,6 +613,18 @@ suite('navigation >', function() {
         }).then(done, done);
       });
     });
+
+    suite('setReady', function() {
```
Comment (steveck-chung) [inline apps/sms/views/shared/test/unit/navigation_test.js]: This trick is only for non-split view, so only add it in the non-split test suite.

File: apps/sms/views/shared/test/unit/sms_test.js [hunk 6]
```diff
@@ -326,6 +326,13 @@ suite('SMS App Unit-Test', function() {
         assertNumberOfElementsInContainerByTag(container, 5, 'input');
       });
 
+      test('Sets up the gaia header for the edit form', function() {
```
Comment (steveck-chung) [inline apps/sms/views/shared/test/unit/sms_test.js]: The selection related tests does not live in the inbox/conversation unit test but sms test for a long time. Adding the tests here for now but we might need to move them back to view specific unit test.

**Comment by julienw** [comment]
this needs a simple rebase

**Comment by steveck-chung** [comment]
Rebased!

File: apps/sms/views/conversation/test/unit/conversation_test.js [hunk 7]
```diff
@@ -6681,7 +6683,10 @@ suite('conversation.js >', function() {
 
       test('enables the font-fit algorithm in headers', function() {
         assert.isFalse(header.hasAttribute('no-font-fit'));
-        assert.isFalse(editHeader.hasAttribute('no-font-fit'));
```
Comment (julienw) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: mmm this test should move in the edit mode tests, I think ?

File: apps/sms/views/inbox/test/unit/inbox_test.js [hunk 8]
```diff
@@ -1960,13 +1960,6 @@ suite('thread_list_ui', function() {
 
       assert.isFalse(draftSavedBanner.classList.contains('hide'));
     });
-
-    test('Sets up the gaia header for the edit form', function() {
-      var editHeader = document.getElementById('threads-edit-header');
-      assert.isTrue(editHeader.hasAttribute('no-font-fit'));
-      InboxView.beforeEnter();
-      assert.isFalse(editHeader.hasAttribute('no-font-fit'));
-    });
```
Comment (julienw) [inline apps/sms/views/inbox/test/unit/inbox_test.js]: same here

File: apps/sms/views/shared/js/startup.js [hunk 9]
```diff
@@ -79,8 +79,24 @@ var Startup = exports.Startup = {
...
+    '/shared/style/input_areas.css',
+    '/shared/style/progress_activity.css',
+    '/views/shared/style/composer.css',
+    '/views/conversation/style/message.css',
+    '/views/conversation/style/report_view.css',
+    '/views/conversation/style/attachment.css',
+    '/views/conversation/style/conversation.css',
+    '/views/new_message/style/new_message.css',
+    '/views/new_message/style/recipients.css'
+  ],
```
Comment (julienw) [inline apps/sms/views/shared/js/startup.js]: wondering about lazyloading these in ConversationView and InformationView beforeEnter methods :) but maybe for later

File: apps/sms/views/shared/test/unit/navigation_test.js [hunk 10]
```diff
@@ -611,6 +613,18 @@ suite('navigation >', function() {
...
+
+      test('Remove hidden class once the navigation is ready', function() {
+
+        assert.isTrue(document.querySelectorAll('.panel-hidden').length > 0);
```
Comment (julienw) [inline apps/sms/views/shared/test/unit/navigation_test.js]: nit: I don't know if we have it in our chai: `assert.isAbove`

File: apps/sms/views/shared/test/unit/navigation_test.js [hunk 11]
```diff
@@ -611,6 +613,18 @@ suite('navigation >', function() {
...
+
+        Navigation.setReady();
+
+        assert.equal(document.querySelectorAll('.panel-hidden').length, 0);
```
Comment (julienw) [inline apps/sms/views/shared/test/unit/navigation_test.js]: and here: assert.lengthOf

File: apps/sms/views/shared/test/unit/sms_test.js (hunk 6 above)
Comment (julienw) [inline apps/sms/views/shared/test/unit/sms_test.js]: mmm I see; we have no edit mode tests elsewhere ? can you file a bug for this ?

File: apps/sms/views/conversation/test/unit/conversation_test.js (hunk 7 above)
Comment (julienw) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: ok I see it's in sms_test...

File: apps/sms/views/inbox/test/unit/inbox_test.js (hunk 8 above)
Comment (julienw) [inline apps/sms/views/inbox/test/unit/inbox_test.js]: same

File: apps/sms/views/shared/js/startup.js (hunk 9 above)
Comment (steveck-chung) [inline apps/sms/views/shared/js/startup.js]: I think most of the lazy load part would be disappeared naturally once we introduce the real split view.

File: apps/sms/views/shared/test/unit/navigation_test.js (hunk 10 above)
Comment (steveck-chung) [inline apps/sms/views/shared/test/unit/navigation_test.js]: We don't have the `isAbove`, but `lengthOf` is existed and in use.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=428a8f645fe425662ed7bd78455d061ebf3d3692

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f1c836923010ce6e113cdd06c1ea29e1c792f68e

**Pull request merged**